0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,180 @@
+https://3.basecamp.com/4403044/buckets/16118249/messages/2583518881#__recording_2600021433[in basecamp].
+
+The final model can be found at the end of this document.
+Our reasoning for creating certain table relations is summed up in the sections below.
+
+=== Swaps
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/swap.puml&fmt=svg[Swaps]
+
+Our first challange was depicting `alpha` and `beta` protocols for the swap.
+Our first attempt was to not indicate this in the `swap` table but instead just use foreign key relationships.
+For that to work we would need to indicate in the child table whether it is `alpha` or `beta` (see below in `swap_protocol`).
+In order to reduce the amount of duplicated data we decided to add a separate table `protocol` and a `1 to many` relationship from `swap` to `protocol`.
+`protocol` is a _static_ table.
+Static means, it contains all available protocols as identifiers.
+For now: `han_bitcoin`, `han_ethereum`, `herc20`, `halight`.
+
+This allows us to :
+
+* simple update the name if we decide to, i.e. instead of updating X rows in `swaps` we update 1 entry in `protocol`
+* add additional protocols
+
+The second thing which was a bit challenging was how to model data which is only known at a later point in time and how to model data which not all `Swaps` have in common.
+We decided to go for an _extension table_ approach which allows us to add additional information where needed:
+
+* Bob only learns the hash of the secret during the announcement, additionally not every Swap has a secret hash.
+* The shared swap id is only available after a swap was finalized.
+
+We added `secret_hash` and `shared_swap_id` as extension tables, that have the `swap_local_id` as primary key (so there can only be one entry per swap).
+This allows us to add information to one specific swap without having to update the `swap` table.
+Note that the actual secret is not stored in the database, Alice always computes it from the swap parameters and her seed.
+
+Third we dealt with the problem that there is not always an `address_hint` (i.e. Bob does not need to know the `address_hint` of Alice).
+We added the `address_hint` extension table that relates to `swap` on the `counter_party_peer_id` field (rather than on the `swap_local_id`).
+
+Note that the role could be factored out of the table, but we decided against it because it adds unnecessary complexity for now.
+
+What remains is a simple table for `swap` which contains only the information which all swaps have in common: `role`, `alpha_protocol`, `beta_protocol` and `counter_party_peer_id`.
+
+=== Protocols
+
+[[HanProtocols]]
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/han_swap_protocols.puml&fmt=svg[HanProtocols]
```

Comment:
I am not convinced about the `han-bitcoin_swap_Detial_id` and other foreign keys, what not using the swap_local_id as a foreign and primary key?

Reply:
Why having the identities in separate tables?

Reply:
Why having the state in a separate table?

Reply:
> I am not convinced about the han-bitcoin_swap_Detial_id and other foreign keys, what not using the swap_local_id as a foreign and primary key?

It's mentioned a few lines below: 

```
* `han_bitcoin/ethereum_swap_detail` contains data needed to create the HTLCs. We decided to give this table a generated primary key, rather than using `swap_local_id` as the primary key to allow same chain swaps.
```

> Why having the identities in separate tables?

5 lines below: 

```
* `han_bitcoin_ethereum_redeem/refund_identity`: Since some identities are exchanged during announcement, the identities are kept in extension tables.
```

> Why having the state in a separate table?

another 2 lines further down: 

```
han_bitcoin/ethereum_ledger_state: is a new table which we introduced in this database schema. It records observed ledger events such as fund/redeem/refund-transaction. We extracted the states into han_state in order to reduce duplication in each table and not have an extra column for the event-type (fund/redeem/refund). This allows us to reuse these states among all han_* protocols.
```